trie/sync.go

```diff
@@ -418,25 +467,70 @@ func (s *Sync) children(req *request, object node) ([]*request, error) {
 // commit finalizes a retrieval request and stores it into the membatch. If any
 // of the referencing parent requests complete due to this commit, they are also
 // committed themselves.
-func (s *Sync) commit(req *request) (err error) {
+func (s *Sync) commitNodeRequest(req *nodeRequest) error {
 	// Write the node content to the membatch
-	if req.code {
-		s.membatch.codes[req.hash] = req.data
-		delete(s.codeReqs, req.hash)
-		s.fetches[len(req.path)]--
-	} else {
-		s.membatch.nodes[req.hash] = req.data
-		delete(s.nodeReqs, req.hash)
-		s.fetches[len(req.path)]--
+	s.membatch.nodes[req.key] = req.data
+	delete(s.nodeReqs, req.key)
+	s.fetches[len(req.path)]--
+
+	// Check all parents for completion
+	for _, parent := range req.parents {
+		parent.deps--
+		if parent.deps == 0 {
+			if err := s.commitNodeRequest(parent); err != nil {
+				return err
+			}
+		}
 	}
+	return nil
+}
+
+// commit finalizes a retrieval request and stores it into the membatch. If any
+// of the referencing parent requests complete due to this commit, they are also
+// committed themselves.
+func (s *Sync) commitCodeRequest(req *codeRequest) error {
+	// Write the node content to the membatch
+	s.membatch.codes[req.hash] = req.data
+	delete(s.codeReqs, req.hash)
+	s.fetches[len(req.path)]--
+
 	// Check all parents for completion
 	for _, parent := range req.parents {
 		parent.deps--
 		if parent.deps == 0 {
-			if err := s.commit(parent); err != nil {
+			if err := s.commitNodeRequest(parent); err != nil {
 				return err
 			}
 		}
 	}
 	return nil
 }
+
+// encodeKey generates a unique key for the trie node with the given
+// path and hash. The path also contains the path in parent trie if
+// it's a two-layered trie node.
+func encodeKey(path []byte, hash common.Hash) string {
```

Comment:
I *think* the hash is not needed here since everything is uniquely identified by path, and in that case, the key just becomes the original path, so that could be dropped too.